pkg/params/params.go

```diff
@@ -0,0 +1,240 @@
+}
+
+// PreValidate validates a string against the given parameter
+func (p *Param) PreValidate(value string) error {
+	if value == "" && p.IsMandatory {
+		return fmt.Errorf("expected value for %q", p.Key)
+	}
+
+	if len(p.PossibleValues) > 0 {
+		for _, v := range p.PossibleValues {
+			if v == value {
+				return nil
+			}
+		}
+		return fmt.Errorf("invalid value %q as %q: valid values are: %s", value, p.Key, strings.Join(p.PossibleValues, ", "))
+	}
+	if validator, ok := typeHintValidators[p.TypeHint]; ok {
+		if err := validator(value); err != nil {
+			return fmt.Errorf("invalid value %q as %q: %w", value, p.Key, err)
+		}
+	}
+	if p.Validator != nil {
+		if err := p.Validator(value); err != nil {
+			return fmt.Errorf("invalid value %q as %q: %w", value, p.Key, err)
+		}
+	}
+
+	return nil
+}
+
+// Validate validates an already set value
+func (p *Param) Validate() error {
+	return p.PreValidate(p.String())
+}
+
+func (p *Params) AddParams(other Params) {
+	for _, v := range other {
+		*p = append(*p, v)
+	}
+}
+
+func (p *Params) AddParam(key, value string) {
+	*p = append(*p, &Param{
+		Key:      key,
+		value:    value,
+		assigned: true,
+	})
+}
+
+func (p Params) Set(key, val string) error {
+	for _, e := range p {
+		if e.Key == key {
+			return e.Set(val)
+		}
+	}
+	return fmt.Errorf("param %q not found", key)
+}
+
+func (p *Params) ParamMap() (res ParamMap) {
+	res = make(ParamMap)
+	for _, v := range *p {
+		res[v.Key] = v.String()
+	}
+	return
+}
+
+func (p *Params) Validate() error {
+	for _, param := range *p {
+		if err := param.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Get returns the value of the parameter with the given key or an empty string
+func (p *Params) Get(key string) string {
+	for _, param := range *p {
+		if key == param.Key {
+			return param.String()
+		}
+	}
+	return ""
+}
+
+func (p *Params) ValidateStringMap(cfg map[string]string) error {
+	for _, param := range *p {
+		value, ok := cfg[param.Key]
+		if !ok && param.IsMandatory {
+			return fmt.Errorf("expected value for %q", param.Key)
+		}
+		if param.Validator != nil {
+			if err := param.Validator(value); err != nil {
+				return fmt.Errorf("invalid value %q as %q: %w", value, param.Key, err)
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Param) String() string {
+	if p == nil {
+		return ""
+	}
+	if p.assigned {
+		return p.value
+	}
+	return p.DefaultValue
+}
+
+func (p *Param) Set(val string) error {
+	if p.Validator != nil {
+		err := p.Validator(val)
+		if err != nil {
+			return err
+		}
+	}
+	p.assigned = true
+	p.value = val
+	return nil
+}
+
+func (p *Param) Type() string {
+	if p.TypeHint == TypeBool {
+		return "bool"
+	}
+	return "string"
+}
+
+func (p *Param) IsBoolFlag() bool {
+	return p.TypeHint == TypeBool
+}
+
+func ParamMapFromParams(p Params) map[string]*Param {
+	res := make(map[string]*Param)
+	for _, param := range p {
+		res[param.Key] = param
+	}
+	return res
+}
```

Comment:
Not needed after my previous comment? 

(Please notice that this function is a bit confusing as is name states "ParamMap" but returns a different type than `ParamMap`)